### PR TITLE
PowerShell transcripts should include the configuration name in the transcript header

### PR DESCRIPTION
**Pull request by chunqingchen**
Steps to reproduce
```powershell
New-PSSessionConfigurationFile -Path .\myJeaConfig.pssc -TranscriptDirectory 'C:\temp\transcripts' -SessionType RestrictedRemoteServer
Register-PSSessionConfiguration -Name JEA -Path .\myJeaconfig.pssc
Enter-PSSession -ComputerName Localhost -ConfigurationName JEA

exit
```
Expected behavior

PowerShell transcripts currently do not log the configuration name the user used to connect to and manage the machine. For JEA scenarios, this means an auditor trying to understand how someone was able to do a certain command will not know through which endpoint the user entered and was assigned those privileges.

Suggestion is to add a new line to the transcript header similar to the following:

ConfigurationName: MyJEA

[…]
RunAs User: WinRM Virtual Users\WinRM VA_10_PRIV_priv.demo
Configuration Name: JEA
Machine: DC (Microsoft Windows NT 10.0.14393.0)
[…]

Actual behavior

[…]
RunAs User: WinRM Virtual Users\WinRM VA_10_PRIV_priv.demo
Machine: DC (Microsoft Windows NT 10.0.14393.0)
[…]

Environment data
```
Name Value

PSVersion 5.1.14993.1000
PSEdition Desktop
PSCompatibleVersions {1.0, 2.0, 3.0, 4.0...}
BuildVersion 10.0.14993.1000
CLRVersion 4.0.30319.42000
WSManStackVersion 3.0
PSRemotingProtocolVersion 2.3
SerializationVersion 1.1.0.1
```

### CODE REVIEW

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs]: This may be more efficient for extracting the configuration name from the Shell Uri since it only allocates one string instead of multiple string objects:

```
string shellPrefix = System.Management.Automation.Remoting.Client.WSManNativeApi.ResourceURIPrefix;
int index = shelluri.IndexOf(shellPrefix, StringComparison.OrdinalIgnoreCase);
return (index == 0) ? shelluri.Substring(shellPrefix.Length) : string.Empty;
```

File: src/System.Management.Automation/engine/remoting/server/serverremotesession.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/remoting/server/serverremotesession.cs]: It would be better to parse the configuration name from the shellUri (configurationProcviderId) here rather than parse it each time later when a transcript header is written.

**Review by PaulHigin — DISMISSED** [review]

File: src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs
Comment (lzybkr) [inline src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs]: You should prefer auto-properties where possible.

**Comment by chunqingchen** [comment]
Thanks @lzybkr, @PaulHigin  ! I have incorporated the suggested change.

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs]: >psConfigurationName [](start = 20, length = 19)

I believe that this cannot be null.  Please make it default to empty string.

**Comment by chunqingchen** [comment]
@PaulHigin the request has been addressed.

**Comment by mirichmo** [comment]
@chunqingchen Please include tests in this PR

**Review by SteveL-MSFT — DISMISSED** [review]
test needs to be added

**Comment by mirichmo** [comment]
@chunqingchen You need to add yourself to Microsoft on GitHub

**Comment by joeyaiello** [comment]
@chunqingchen I suspect you're already in the orgs, you just need to do this (for both PowerShell and Microsoft, though doing Microsoft is what will fix the CLA bot): https://help.github.com/articles/publicizing-or-hiding-organization-membership/

**Comment by TravisEz13** [comment]
There are open comments that still need to be addressed.  See: https://github.com/PowerShell/PowerShell/pull/2890#issuecomment-271920381

**Comment by TravisEz13** [comment]
We have been waiting on author response to PR comments for over a week.

**Comment by chunqingchen** [comment]
working on the test now

**Comment by TravisEz13** [comment]
@chunqingchen Please don't dismiss (delete) other people reviews.  If you believe someone review is bad enough to be dismissed, please talk to a maintainer.

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs
Comment (TravisEz13) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs]: @PaulHigin , I belive this is resolved.

**Review by TravisEz13 — DISMISSED** [review]

**Comment by chunqingchen** [comment]
@TravisEz13 I meant the reviews are resolved. So dismiss is not the way to say it has been resolved. Got it.

**Comment by chunqingchen** [comment]
@PaulHigin Hi Paul, I talked with Travis and he agrees to merge the commit if you are fine. do you have any more comment about this change?

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (iSazonov) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Please use more informative title.
Sample - "Transcript header".

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (iSazonov) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Please add new line.

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (iSazonov) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Please replace `-force` with `-Force`.

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (iSazonov) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Please replace aliases with full cmdlet names.

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (iSazonov) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Please use `Should BeLike` or `Should BeLikeExactly`

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (iSazonov) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Please replace `-force` with `-Force`.

**Comment by PaulHigin** [comment]
I agree with @iSazonov comments.  Otherwise LGTM.

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
I agree with @iSazonov comments. Otherwise LGTM.

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (TravisEz13) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: This is recommended, but not required.

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (TravisEz13) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: Yes, please replace `bug fix test` with something more descriptive.

Comment (chunqingchen) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: corrected

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (chunqingchen) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: corrected

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (chunqingchen) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: corrected

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (chunqingchen) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: corrected

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (chunqingchen) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: corrected

**Comment by chunqingchen** [comment]
@iSazonov @TravisEz13 comments are all resolved

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1
Comment (TravisEz13) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1]: This was not fixed

**Review by TravisEz13 — APPROVED** [review]

**Comment by iSazonov** [comment]
Thanks!
Tests LGTM.

**Comment by jpsnover** [comment]
Great change!

**Pull request merged**
